pennylane/transforms/decorators.py

```diff
@@ -0,0 +1,388 @@
+    ...     qml.Hadamard(wires=0)
+    ...     qml.CRX(-0.5, wires=[0, 1])
+    >>> new_tape = my_transform(tape, 1., 2.)
+    >>> print(new_tape.draw())
+     0: ──H───────────────╭Z──┤
+     1: ──RX(0.5)──RY(1)──╰C──┤
+    """
+
+    def __init__(self, transform_fn):
+        self.transform_fn = transform_fn
+        functools.update_wrapper(self, transform_fn)
+
+    def __call__(self, tape, *args, **kwargs):
+        tape_class = type(tape.__class__.__name__, (NonQueuingTape, tape.__class__), {})
+
+        with tape_class() as new_tape:
+            self.transform_fn(tape, *args, **kwargs)
+
+        new_tape.__class__ = tape.__class__
+        return new_tape
+
+
+def qfunc_transform(tape_transform):
+    """Converts a single tape transform to a quantum function (qfunc) transform.
+
+    Args:
+        tape_transform (single_tape_transform): the single tape transform
+            to convert into the qfunc transform.
+
+    Returns:
+        function: A qfunc transform, that acts on any qfunc returning a *new*
+        qfunc as per the tape transform.
+
+    **Example**
+
+    Given a single tape transform ``my_transform(tape, x, y)``, you can use
+    this function to convert it into a qfunc transform:
+
+    >>> my_qfunc_transform = qfunc_transform(my_transform)
+
+    It can then be used to transform an existing qfunc:
+
+    >>> new_qfunc = my_qfunc_transform(old_qfunc, 0.6, 0.7)
+
+    It can also be used as a decorator:
+
+    .. code-block:: python
+
+        @qml.qfunc_transform
+        @qml.single_tape_transform
+        def my_transform(tape, x, y):
+            for op in tape.operations + tape.measurements:
+                if op.name == "CRX":
+                    wires = op.wires
+                    param = op.parameters[0]
+                    qml.RX(x * param, wires=wires[1])
+                    qml.RY(y * qml.math.sqrt(param), wires=wires[1])
+                    qml.CZ(wires=[wires[1], wires[0]])
+                else:
+                    op.queue()
+
+        @my_transform(0.6, 0.1)
+        def qfunc(x):
+            qml.Hadamard(wires=0)
+            qml.CRX(x, wires=[0, 1])
+
+    Let's use this qfunc to create a QNode, so that we can execute it on a quantum
+    device:
+
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> qnode = qml.QNode(qfunc, dev)
+    >>> print(qml.draw(qnode)(2.5))
+     0: ──H───────────────────╭Z──┤
+     1: ──RX(1.5)──RY(0.158)──╰C──┤
+
+    The not only is the transformed qfunc fully differentiable, but the qfunc transform
+    parameters *themselves* are differentiable:
+
+    .. code-block:: python
+
+        dev = qml.device("default.qubit", wires=2)
+
+        def ansatz(x):
+            qml.Hadamard(wires=0)
+            qml.CRX(x, wires=[0, 1])
+
+        @qml.qnode(dev)
+        def circuit(param, transform_weights):
+            qml.RX(0.1, wires=0)
+
+            # apply the transform to the ansatz
+            my_transform(*transform_weights)(ansatz)(param)

```

Comment:
Oh this is really cool! And I think this clears up my not understanding how `ctrl` and `adjoint` fit in this category; it looks like `my_transform` is being used here in a similar way?

Reply:
yep! exactly 🙂 Both `adjoint` and `ctrl` don't have any 'transform parameters', so they simply look like:
```python
qml.adjoint(ansatz)(param)
```
But since this one has additional (trainable) transform parameters, you have that extra level of nesting.

Reply:
Oh actually, I think this is an issue. I've coded this up for the most general case, which means _no_ transform parameters would look like this:

```python
my_transform()(ansatz)(param)
```

🙁 

I'm not 100% sure it's possible to have a decorator that supports both cases (transform params + no transform params).

Reply:
So, this is actually a Python limitation as far as I can tell; you can either create a decorator that has parameters, or does not, you can't have it change dynamically.

So we have two options:

- It always permits parameters, which results in the ugly `my_transform()(ansatz)(params)` if your transform has no parameters.
- Alternatively, we could add _another_ level of nesting to the decorator, and use `inspect.signature` to determine the decorator wrapping. This _should_ work, but haven't tested it.

Reply:
I think I just ran into this very problem while working with a tape/qfunc transform that didn't have any parameters. I had to construct the decorator like `@transform()` to get a QNode I used it in to actually run. It would be nice to just not have to worry about it - if we were to add another level of nesting, as you suggest, that would make both cases possible without the user having to worry about it? (From a user standpoint I feel like that's the best option, if it works)

Reply:
@glassnotes I implemented option (2) above using `inspect.signature`, it works really nicely :)

```python

@qml.qfunc_transform
@qml.single_tape_transform
def convert_cnots(tape):
    """qfunc transform with no arguments"""
    for op in tape.operations + tape.measurements:
        if op.name == "CNOT":
            wires = op.wires
            qml.Hadamard(wires=wires[1])
            qml.CZ(wires=[wires[1], wires[0]])
            qml.Hadamard(wires=wires[1])
        else:
            op.queue()

@qml.qfunc_transform
@qml.single_tape_transform
def expand_hadamards(tape, x, y):
    """qfunc transform with arguments"""
    for op in tape.operations + tape.measurements:
        if op.name == "Hadamard":
            qml.RZ(x, wires=op.wires)
            qml.RY(y, wires=op.wires)
        else:
            op.queue()

dev = qml.device("default.qubit", wires=3)

@expand_hadamards(0.5, 0.9)  # pass arguments here
@convert_cnots  # no arguments, so no parenthesis
def ansatz():
    qml.CNOT(wires=[0, 1])
    qml.CNOT(wires=[1, 2])
    qml.CNOT(wires=[2, 0])
```

Or, alternatively:

```python
def ansatz():
    qml.CNOT(wires=[0, 1])
    qml.CNOT(wires=[1, 2])
    qml.CNOT(wires=[2, 0])

transformed_ansatz = expand_hadamards(0.5, 0.1)(convert_cnots(ansatz))

@qml.qnode(dev)
@expand_hadamards(1., 2.)
def my_other_circuit(x):
    qml.RX(x, wires=0)
    transformed_ansatz()
    qml.Hadamard(wires=2)
    return qml.expval(qml.PauliZ(0))
```

If you could test it out and see if it fixes it on your end, that would be great!